miniapps/adjoint/advection_diffusion.cpp

```diff
@@ -0,0 +1,502 @@
+   }
+
+   // Set up material properties, and primal and adjoint variables
+   // p are the fixed material properties
+   Vector p(2);
+   p[0] = 1.0;
+   p[1] = 0.5;
+
+   // U is the size of the solution/primal vector
+   // Set U with the initial conditions
+   ParGridFunction u(fes);
+   FunctionCoefficient u0(u_init);
+   u.ProjectCoefficient(u0);
+
+   cout << "Init u: " << endl;
+   u.Print();
+
+   // TimeDependentOperators need to be TrueDOF Size
+   HypreParVector *U = u.GetTrueDofs();
+
+   // Get boundary conditions
+   Array<int> ess_tdof_list;
+   Array<int> essential_attr(pmesh->bdr_attributes.Size());
+   essential_attr[0] = 1;
+   essential_attr[1] = 1;
+   fes->GetEssentialTrueDofs(essential_attr, ess_tdof_list);
+
+   // Setup the TimeDependentAdjointOperator and the CVODESSolver  
+   AdvDiffSUNDIALS adv(U->Size(), U->Size(), p, fes, ess_tdof_list);
+
+   // Set the initial time to the TimeDependentAdjointOperator
+   double t = 0.0;
+   adv.SetTime(t);
+
+   // Create the CVODES solver corresponding to the selected step method
+   CVODESSolver *cvodes = new CVODESSolver(fes->GetComm(),
+					   step_mode ? CV_ADAMS : CV_BDF);
+   cvodes->Init(adv);
+   cvodes->UseSundialsLinearSolver();
+
+   // Relative and absolute tolerances for CVODES
+   double reltol = 1e-8, abstol = 1e-6;
+   cvodes->SetSStolerances(reltol, abstol);
+
+   // Initialize adjoint problem settings
+   int checkpoint_steps = 50; // steps between checkpoints
+   cvodes->InitAdjointSolve(checkpoint_steps, CV_HERMITE);
+
+   // Perform time-integration for the problem (looping over the time
+   // iterations, ti,  with a time-step dt).
+   bool done = false;
+   for (int ti = 0; !done; )
+   {
+      double dt_real = max(dt, t_final - t);
+      cvodes->Step(*U, t, dt_real);
+      ti++;
+
+      done = (t >= t_final - 1e-8*dt);
+
+      if (done )
+      {
+         cvodes->PrintInfo(); 
+      }
+   }
+
+   u = *U;
+   cout << "Final Solution: " << t << endl;
+   u.Print();
+
+   // Calculate the quadrature int_x u dx at t = 5
+   // Since it's only a spatial quadrature we evaluate it at t=5
+   ParLinearForm obj(fes);
+   ConstantCoefficient one(1.0);
+   obj.AddDomainIntegrator(new DomainLFIntegrator(one));
+   obj.Assemble();
+
+   double g = obj(u);
+   if (myid == 0)
+   {
+      cout << "g: " << g << endl;
+   }
+
+   // Solve the adjoint problem. v is the adjoint solution
+   ParGridFunction v(fes);
+   v = 1.;
+   v.SetSubVector(ess_tdof_list, 0.0);
+   HypreParVector *V = v.GetTrueDofs();
+
+   // Initialize quadrature sesntiviity values to zero
+   Vector qBdot(p.Size());
+   qBdot = 0.;
+   
+   V->Vector::Print();
+
+   t = t_final;
+   cvodes->InitB(adv);
+   cvodes->InitQuadIntegrationB(qBdot, 1.e-6, 1.e-6);
+   cvodes->UseSundialsLinearSolverB();
+   cvodes->SetSStolerancesB(reltol, abstol);
+
+   // Results at time TBout1
+   double dt_real = max(dt, t);
+   cvodes->StepB(*V, t, dt_real);
+   cout << "t: " << t << endl;
```

Comment:
I would print them only on process 0.

Reply:
How about I print time on process 0, so you can compare the results between np = 1 and np = 2 tasks?